File: Stack/MatchingParenthesesSearch/BracketChecker.swift
class Stack {
    private var maxSize: Int
    private var stackArray: [Character]
    private var top: Int = -1

    init(_ s: Int) {
        maxSize = s
        stackArray = Array(repeating: Character(" "), count: maxSize)
        top = -1
    }

    func push(_ j: Character) -> Void {
        top += 1
        stackArray[top] = j
    }

    func pop() -> Character {
        let element = stackArray[top]
        top -= 1
        return element
    }

    func peek() -> Character {
        return stackArray[top]
    }

    func isEmpty() -> Bool {
        return top == -1
    }
}

func checkParentheses(_ input: String) -> Bool {
    let theStack = Stack(input.count)

    for ch in input {
        if ch == "(" || ch == "{" || ch == "[" {
            theStack.push(ch)
        } else if ch == ")" || ch == "}" || ch == "]" {
            if theStack.isEmpty() {
                return false
            }

            let top = theStack.pop()
            if (ch == ")" && top != "(") ||
               (ch == "}" && top != "{") ||
               (ch == "]" && top != "[") {
                return false
               }
        }
    }

    return theStack.isEmpty()
}

let input = "{(2+3)*[5-2]}"
let isBalanced = checkParentheses(input)

if isBalanced {
    print("Parentheses are balanced")
} else {
    print("Parentheses are not balanced")
}

File: Stack/MatchingParenthesesSearch/BracketChecker.go
package main

import "fmt"

type Stack struct {
	maxSize    int
	stackArray []rune
	top        int
}

func NewStack(s int) *Stack {
	return &Stack{
		maxSize:    s,
		stackArray: make([]rune, s),
		top:        -1,
	}
}

func (s *Stack) push(j rune) {
	s.top++
	s.stackArray[s.top] = j
}

func (s *Stack) pop() rune {
	element := s.stackArray[s.top]
	s.top--
	return element
}

func (s *Stack) peek() rune {
	return s.stackArray[s.top]
}

func (s *Stack) isEmpty() bool {
	return s.top == -1
}

func checkParentheses(input string) bool {
	theStack := NewStack(len(input))

	for _, ch := range input {
		if ch == '(' || ch == '{' || ch == '[' {
			theStack.push(ch)
		} else if ch == ')' || ch == '}' || ch == ']' {
			if theStack.isEmpty() {
				return false
			}

			top := theStack.pop()
			if (ch == ')' && top != '(') ||
				(ch == '}' && top != '{') ||
				(ch == ']' && top != '[') {
				return false
			}
		}
	}

	return theStack.isEmpty()
}
func main() {
	input := "{(2+3)*[5-2]}"
	isBalanced := checkParentheses(input)

	if isBalanced {
		fmt.Println("Parentheses are balanced")
	} else {
		fmt.Println("Parentheses are not balanced")
	}
}